redis/cluster/redis.go

```diff
@@ -0,0 +1,182 @@
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/redis/go-redis/extra/redisotel/v9"
+	"github.com/redis/go-redis/v9"
+	"github.com/spf13/viper"
+)
+
+// Client identifies uniquely one redis client with a pool of connections
+type Client struct {
+	Instance    redis.UniversalClient
+	ClusterMode bool
+}
+
+type ClientArgs struct {
+	Url           string
+	ClusterMode   bool
+	EnableMetrics bool
+	EnableTracing bool
+}
+
+// NewClient creates and returns a new redis client based on the given settings. It only supports redis 7 engine and uses go-redis v9.
+func NewClient(args *ClientArgs) (*Client, error) {
+	if args == nil {
+		return nil, fmt.Errorf("NewClient must have a non-nil args")
+	}
+
+	client := &Client{
+		ClusterMode: args.ClusterMode,
+	}
+
+	if args.ClusterMode {
+		if err := client.ConnectCluster(args.Url); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := client.Connect(args.Url); err != nil {
+			return nil, err
+		}
+	}
+
+	if args.EnableTracing {
+		if err := redisotel.InstrumentTracing(client.Instance); err != nil {
+			return nil, err
+		}
+	}
+
+	if args.EnableMetrics {
+		if err := redisotel.InstrumentMetrics(client.Instance); err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+// NewClientFromConfig creates and returns a new redis client based on the given settings from predefined env vars. It only supports redis 7 engine and uses go-redis v9.
+func NewClientFromConfig(config *viper.Viper, prefix string) (*Client, error) {
+	if config == nil {
+		return nil, fmt.Errorf("NewClientFromConfig must have a non-nil config")
+	}
+
+	args := CreateClientArgs(config, prefix)
+
+	return NewClient(args)
+}
+
+func CreateClientArgs(config *viper.Viper, prefix string) *ClientArgs {
+	url := config.GetString(fmt.Sprintf("%s.url", prefix))
+
+	if url == "" {
+		tls := config.GetBool(fmt.Sprintf("%s.tls", prefix))
+
+		endpoint := config.GetString(fmt.Sprintf("%s.endpoint", prefix))
+		port := config.GetString(fmt.Sprintf("%s.port", prefix))
+		pass := config.GetString(fmt.Sprintf("%s.password", prefix))
+
+		var baseUrl string
+		if tls {
+			baseUrl = "rediss://"
+		} else {
+			baseUrl = "redis://"
+		}
+
+		url = fmt.Sprintf("%s:%s@%s:%s", baseUrl, pass, endpoint, port)
+	}
+
+	clusterMode := config.GetBool(fmt.Sprintf("%s.clusterMode", prefix))
+	enableMetrics := config.GetBool(fmt.Sprintf("%s.enableMetrics", prefix))
+	enableTracing := config.GetBool(fmt.Sprintf("%s.enableTracing", prefix))
+
+	// Extra parameters
+	var queryParameters []string
+	maxRedirects := config.GetString(fmt.Sprintf("%s.maxRedirects", prefix))
+	if maxRedirects != "" {
+		queryParameters = append(queryParameters, fmt.Sprintf("max_redirects=%s", maxRedirects))
+	}
+
+	poolSize := config.GetString(fmt.Sprintf("%s.poolSize", prefix))
+	if poolSize != "" {
+		queryParameters = append(queryParameters, fmt.Sprintf("pool_size=%s", poolSize))
+	}
+
+	readTimeout := config.GetString(fmt.Sprintf("%s.timeout", prefix))
```

Comment:
Besides `readTimeout` I'd recommend supporting `dialTimeout` (connection timeout) and `writeTimeout` (socket writes timeout) https://github.com/redis/go-redis/blob/v9.7.1/options.go#L31

In previous versions of the lib we had a custom connection timeout as the connection "waiting time" wasn't available in the API (I guess), now that it is we should take advantage of